Sming/Libraries/Servo/Servo.cpp

```diff
@@ -7,79 +7,77 @@
 
 #include "Servo.h"
 
-#include <HardwareTimer.h>
-
 // DEBUG for visual check of impulses with a LED instead of the servo
 //#define DEBUG
 #ifdef DEBUG
 #define FACTOR 500
+#else
+#define FACTOR 1
 #endif
 
 #define SERVO_PERIOD 20000
 
+HardwareTimer Servo::hardwareTimer;
+uint8 Servo::pins[SERVO_CHANNEL_NUM_MAX];
+uint32 Servo::timing[SERVO_CHANNEL_NUM_MAX * 2 + 1];
+uint8 Servo::maxTimingIdx = 0;
+uint8 Servo::actIndex = 0;
+bool Servo::started = false;
 
-HardwareTimer hardwareTimer;
-
-uint8 pins[SERVO_CHANNEL_NUM_MAX];
-uint32 timing[SERVO_CHANNEL_NUM_MAX*2+1];
-uint8 maxTimingIdx;
-uint8 actIndex;
-bool started = false;
-
-void IRAM_ATTR ServoTimerInt()
+void IRAM_ATTR Servo::timerInt()
 {
-	assert(started);
-	hardwareTimer.setIntervalUs(timing[actIndex]);
+	hardwareTimer.setInterval(timing[actIndex]);
 	hardwareTimer.startOnce();
 
-	if (actIndex < maxTimingIdx) {
-		bool out = !(actIndex%2);
-		digitalWrite(pins[actIndex/2],out);
+	if(actIndex < maxTimingIdx) {
+		bool out = !(actIndex % 2);
+		digitalWrite(pins[actIndex / 2], out);
 	}
 
-	if (++actIndex > maxTimingIdx) actIndex = 0;
-}
-
-Servo::Servo()
-{
-	for(uint8 i=0; i<SERVO_CHANNEL_NUM_MAX*2+1; i++) timing[i] = 1000;
-	maxTimingIdx=0;
+	++actIndex;
+	if(actIndex > maxTimingIdx) {
+		actIndex = 0;
+	}
 }
 
-Servo::~Servo()
+Servo::Servo() : channels(SERVO_CHANNEL_NUM_MAX, 0)
 {
-	channels.removeAllElements();
+	for(unsigned i = 0; i < ARRAY_SIZE(timing); i++) {
+		timing[i] = hardwareTimer.usToTicks<1000>();
+	}
 }
 
 bool Servo::addChannel(ServoChannel* channel)
 {
 	uint8 channel_count = channels.size();
-	if (channel_count > SERVO_CHANNEL_NUM_MAX) return false;
+	if(channel_count >= SERVO_CHANNEL_NUM_MAX) {
+		return false;
+	}
 	channels.add(channel);
 
 	ETS_INTR_LOCK();
 	getPins();
 	calcTiming();
 	ETS_INTR_UNLOCK();
```

Comment:
I don't like how this library globally disables interrupts whilst re-calculating values, which can take a (relatively) long time.
The solution which occurs to me is:
1. Double-buffer state information so interrupts only need to be disabled during a pointer swap operation;
2. Introduce an enable/disable Interrupts() method into HardwareTimer.

 I'm inclined to stick a pin in this for now and deal with this separately.

Reply:
Done.